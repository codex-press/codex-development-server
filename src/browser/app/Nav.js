import React from 'react';
import { connect } from 'react-redux'

import { IconInfo, IconEdit, IconSearch, IconConfig } from './icons';
import { toggleModal } from '../actions';
import * as env from '../env';

import './Nav.less';


const mapStateToProps = state => {
  return {
    focus: state.getIn(['ui','focus']),
    article: state.get('article'),
    account: state.get('account'),
  }
}


const mapDispatchToProps = {
  toggleModal,
}



export class Nav extends React.Component {


  constructor(props) {
    super();
    window.addEventListener('keydown', this.windowKey.bind(this));
  }


  render() {
    return (
      <nav className={ 'Nav' + (this.props.focus ? ' focused' : '') } >
        <IconInfo onClick={ () => this.toggleModal('info') } />
        <IconEdit onClick={ () => this.openEdit() } />
        <IconSearch onClick={ () => this.toggleModal('search') } />
        <IconConfig onClick={ () => this.toggleModal('config') } />
      </nav>
    );
  }


  openEdit() {
    let editURL = env.codexOrigin + '/edit' 

    if (this.props.article.get('account_id') === this.props.account.get('id'))
      editURL += '/article' + this.props.article.get('url');

    window.open(editURL, '_blank');
  }


  toggleModal(value) {
    // don't want to let a double click open and then close it
    if (Date.now() - this.lastToggle < 500)
      return;
    this.props.toggleModal(value);
    this.lastToggle = Date.now();
  }


  windowKey(e) {

    const shouldIgnore = (
      e.path[0].tagName === 'INPUT' ||
      e.defaultPrevented ||
      e.metaKey ||
      e.ctrlKey ||
      e.altKey
    );

    if (e.key === 'Escape') {
      e.preventDefault();
      this.props.toggleModal(null);
    }
    else if (shouldIgnore) {
      return;
    }
    else if (e.key === 'f') {
      e.preventDefault();
      this.props.toggleModal('search');
    }
    else if (e.key === 'i') {
      e.preventDefault();
      this.props.toggleModal('info');
    }
    else if (e.key === 'c') {
      e.preventDefault();
      this.props.toggleModal('config');
    }

  }

}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Nav);


