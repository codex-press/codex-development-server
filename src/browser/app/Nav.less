@import (reference) "../index.less";

.Nav {
  position: fixed;
  top: 0;
  left: 50%;
  transform: translateX(-50%);
  padding: 0.1em;

  display: flex;
  flex-flow: row nowrap;

  &:extend(.modal);
  border-top-right-radius: 0;
  border-top-left-radius: 0;
  background: hsl(0, 0%, 31%);

  &:not(.focused) {
    background: hsl(0, 0%, 18%);
  }

  svg {
    cursor: pointer;
    width: 1.4em;
    height: 1.4em;
    fill: currentColor;
    margin: 0.4em;
  }

  svg.info path {
    transform: translateX(50%);
  }

  svg.search path {
    transform: translateY(-5%);
  }

}

