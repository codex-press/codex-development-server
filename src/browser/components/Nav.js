import React from 'react';
import { connect } from 'react-redux'

import { IconInfo, IconEdit, IconSearch, IconConfig } from './icons';
import { toggleModal } from '../actions';

import './Nav.less';


const mapStateToProps = state => {
  return {
    focus: state.getIn(['ui','focus']),
  }
}


const mapDispatchToProps = {
  toggleModal,
}


export class Nav extends React.Component {


  constructor(props) {
    super();
    window.addEventListener('keydown', this.windowKey.bind(this));
  }


  render() {
    return (
      <nav className={ 'Nav' + (this.props.focus ? ' focused' : '') } >
        <IconInfo onClick={ () => this.toggleModal('info') } />
        <IconEdit onClick={ () => this.openEdit() } />
        <IconSearch onClick={ () => this.toggleModal('search') } />
        <IconConfig onClick={ () => this.toggleModal('config') } />
      </nav>
    );
  }


  openEdit() {
    const editUrl = (
      'https://codex.press/edit/article' + window.location.pathname
    );
    window.open(editUrl, '_blank');
  }


  toggleModal(value) {
    // don't want to let a double click open and then close it
    if (Date.now() - this.lastToggle < 500)
      return;
    this.props.toggleModal(value);
    this.lastToggle = Date.now();
  }


  windowKey(e) {

    const shouldIgnore = (
      e.path[0].tagName === 'INPUT' ||
      e.defaultPrevented ||
      e.metaKey ||
      e.ctrlKey ||
      e.altKey
    );

    if (e.key === 'Escape') {
      e.preventDefault();
      this.props.toggleModal(null);
    }
    else if (shouldIgnore) {
      return;
    }
    else if (e.key === 'f') {
      e.preventDefault();
      this.props.toggleModal('search');
    }
    else if (e.key === 'i') {
      e.preventDefault();
      this.props.toggleModal('info');
    }
    else if (e.key === 'c') {
      e.preventDefault();
      this.props.toggleModal('config');
    }

  }

}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Nav);


