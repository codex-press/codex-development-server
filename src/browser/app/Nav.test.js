import React from 'react';
import { shallow } from 'enzyme';
import { Map } from 'immutable';

import { Nav } from './Nav';


test('Renders a Nav', () => {
  expect(shallow(<Nav />)).toMatchSnapshot();
});


test('Clicking the Info icon toggles the modal', () => {
  const toggleModal = jest.fn();
  const c = shallow(<Nav toggleModal={ toggleModal } />);
  c.find('IconInfo').simulate('click');
  expect(toggleModal.mock.calls.length).toBe(1);
  expect(toggleModal.mock.calls[0][0]).toBe('info');
});


test('Clicking the Search icon toggles the modal', () => {
  const toggleModal = jest.fn();
  const c = shallow(<Nav toggleModal={ toggleModal } />);
  c.find('IconSearch').simulate('click');
  expect(toggleModal.mock.calls.length).toBe(1);
  expect(toggleModal.mock.calls[0][0]).toBe('search');
});


test('Clicking the Config icon toggles the modal', () => {
  const toggleModal = jest.fn();
  const c = shallow(<Nav toggleModal={ toggleModal } />);
  c.find('IconConfig').simulate('click');
  expect(toggleModal.mock.calls.length).toBe(1);
  expect(toggleModal.mock.calls[0][0]).toBe('config');
});


test('Clicking the Edit icon opens editor with permission', () => {
  global.open = jest.fn();
  global.location = { pathname: '/hey' };
  const c = shallow(
    <Nav
      account={ Map({ id: 'me' }) }
      article={ Map({ account_id: 'me', url: '/newsy' }) }
    />
  );
  c.find('IconEdit').simulate('click');
  expect(global.open.mock.calls.length).toBe(1);
  let expected = 'https://codex.press/edit/article/newsy';
  expect(global.open.mock.calls[0][0]).toBe(expected);
});


test('Clicking the Edit icon opens editor without permssion', () => {
  global.open = jest.fn();
  global.location = { pathname: '/hey' };
  const c = shallow(
    <Nav account={ Map({ id: 'me' }) } article={ Map({ account_id: 'you' }) } />
  );
  c.find('IconEdit').simulate('click');
  expect(global.open.mock.calls.length).toBe(1);
  expect(global.open.mock.calls[0][0]).toBe('https://codex.press/edit');
});



test('Clicking more than once does not toggle again', () => {
  const toggleModal = jest.fn();
  const c = shallow(<Nav toggleModal={ toggleModal } />);
  c.find('IconInfo').simulate('click');
  c.find('IconInfo').simulate('click');
  c.find('IconInfo').simulate('click');
  c.find('IconInfo').simulate('click');
  expect(toggleModal.mock.calls.length).toBe(1);
});



