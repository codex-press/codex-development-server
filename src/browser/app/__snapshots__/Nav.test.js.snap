// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a Nav 1`] = `
<nav
  className="Nav top"
>
  <IconInfo
    onClick={[Function]}
  />
  <IconEdit
    onClick={[Function]}
  />
  <IconSearch
    onClick={[Function]}
  />
  <IconConfig
    onClick={[Function]}
  />
</nav>
`;
